Add blog post on the Scala 3.3.2 post-mortem. (#1603)
scala/scala-lang: blog/_posts/2024-03-05-scala-3.3.2-post-mortem.md

title: Scala 3.3.2 post-mortem analysis
description: Post-mortem analysis of the Scala 3.3.2 release abort
---

Last week, we announced [the release of Scala 3.4.0 and 3.3.3]({{ site.baseurl }}/blog/2024/02/29/scala-3.4.0-and-3.3.3-released.html).
The announcement mentioned that we had to skip 3.3.2 due to a bug breaking our forward binary [compatibility guarantees](https://docs.scala-lang.org/overviews/core/binary-compatibility-of-scala-releases.html) in a patch release.

In this post, we take a closer look at what happened, why it slipped through our existing testing processes, and what steps we are taking so that it does not happen again in the future.

## Symptoms of the bug

First, let us look at the scenario in which the bug surfaces:

1. Compile a library A with Scala 3.3.2
2. Compile a project B, depending on A, with Scala 3.3.1 (or 3.3.0)
3. The older compiler crashes with the following message:

  ```
  [error] error while loading SomeClass,
  [error] class file .../SomeClass.class is broken, reading aborted with class java.util.NoSuchElementException
  [error] contextual$
  [error] one error found
  ```

The error does not surface in all situations.

> It only happens when `SomeClass`, coming from library A, is involved in the expansion of a macro or inline def.

Reviewer: Actually this happens for any signature at all with a contextual function

Reviewer: ```
echo "object Wrap { def foo = 23; def bar: String ?=> Int = summon[String].length }" | scala-cli compile -d out -S 3.3.2 _.scala

scala-cli run -S 3.3.1 -cp out -e 'println(Wrap.foo)'
```

For example, the name of the class behind an `object Foo` definition has a structured name `ModuleClass("Foo")`, instead of the string `"Foo$"`.

Likewise, generated names for anonymous entites, such as `implicit` parameters, have the structured name `UniqueName("evidence$", 1)`.
The string `"evidence$"` used in the equality definition of names, so that `UniqueName("evidence$", 1)` is not the same name as `UniqueName("foo$", 1)`, but it does not otherwise have any semantic meaning in TASTy.

However, for code readability and performance reasons, the compiler maintains a fixed list of the prefixes it generates, and uses a single instance of the associated `NameKind`, one of its internal data structures.
When it reads the string `"evidence$"`, it reuses the instance of `NameKind` that it uses for that particular string.

This leads us to the cause of our bug.
Between 3.3.1 and 3.3.2, we merged [PR #18280](https://github.com/lampepfl/dotty/pull/18280).
That PR starts with what appeared to be a *refactoring* commit that renamed some `NameKind`s and their associated strings.
Unfortunately, that means the 3.3.2 compiler started emitting `UniqueName`s with the string `"contextual$"`.

> When the 3.3.1 compiler reads that structured name, it tries to find the cached instance of `NameKind` that it should use for `"constextual$"`, and does not find any, because it did not use to generate that string.

Reviewer 2: Typo

```suggestion
When the 3.3.1 compiler reads that structured name, it tries to find the cached instance of `NameKind` that it should use for `"contextual$"`, and does not find any, because it did not use to generate that string.
```

> When the 3.3.1 compiler reads that structured name, it tries to find the cached instance of `NameKind` that it should use for `"contextual$"`, and does not find any, because it did not use to generate that string.

Reviewer: ```suggestion
When the 3.3.1 compiler reads that structured name, it tries to find the cached instance of `NameKind` that it should use for `"contextual$"`, and does not find any, because it did not expect to generate that string.
```

We therefore did not mark the PR as "needs-minor-release" as we merged it into the `main` branch for release in 3.4.0.

Second, the PR was backported to the 3.3.x LTS branch.
Normally, such a refactoring would not have met the bar for *backporting*  on its own.
We backported the commit because the PR that contained it was improving tooling, and in particular performance of incremental compilation.
This is one of the areas for which backports are highly desirable.

Third and last, the backport failed to trigger any Continuous Integration test:

* the initial set of unit tests for the compiler never mix different compiler versions, and
* the community builds compile projects using the *new* compiler with dependencies compiled by the *old* compiler.

However, our issue only manifests when the *old* compiler is used with dependencies compiled by the *new* compiler.

Concretely, we did not have adequate testing scenarios for *forward* compatibility.

The RC cycles failed to surface the issue as well, for a similar reason.
When trying out an RC, users rarely publish the product, even locally, then use it with an older compiler.

> Fortunately, as a last line of defense, the issue was caught by the Play! framework, in-between the release of the 3.3.2 artifacts on Maven Central and the official announcement of the release.

Reviewer 2: ```suggestion
Fortunately, as a last line of defense, the issue was caught by the Play Framework, in-between the release of the 3.3.2 artifacts on Maven Central and the official announcement of the release.
```

Reviewer 2: Maybe reference the GitHub issue?

```suggestion
Fortunately, as a last line of defense, [the issue was caught](https://github.com/playframework/playframework/issues/12418) by the Play Framework in-between the release of the 3.3.2 artifacts on Maven Central and the official announcement of the release.
```

> ## Preventing this kind of bugs in the future

Reviewer 3: ```suggestion
## Preventing this kind of bug in the future
```

As we mentioned in the previous section, the main point of failure was a lack of tests for forward compatibility scenarios.
Our most important fix to the process will therefore be to add adequate forward compatibility tests in the compiler's test suite.

> In addition, this particular bug highlights the importance of accurately following the specificiation of the TASTy format.

Reviewer 4: Specification

> Making additional assumptions on the *origin* of TASTy file---namely that only our own compiler can produce them---leads to problems like this issue.

Reviewer 3: ```suggestion
Making additional assumptions on the *origin* of TASTy files---namely that only our own compiler can produce them---leads to problems like this issue.
```

We will therefore check whether there are other similar assumptions lurking in the compiler's unpickler, and address them.

## Conclusion

> In this post, we examined in detail the bug and events that led to the abandon of the Scala 3.3.2 release.

Reviewer 4: Abandonment? To 3.3.2 being abandoned?

---
layout: blog-detail
post-type: blog
by: Sébastien Doeraene
title: Scala 3.3.2 post-mortem analysis
description: Post-mortem analysis of the Scala 3.3.2 release abort
---

Last week, we announced [the release of Scala 3.4.0 and 3.3.3]({{ site.baseurl }}/blog/2024/02/29/scala-3.4.0-and-3.3.3-released.html).
The announcement mentioned that we had to skip 3.3.2 due to a bug breaking our forward binary [compatibility guarantees](https://docs.scala-lang.org/overviews/core/binary-compatibility-of-scala-releases.html) in a patch release.

In this post, we take a closer look at what happened, why it slipped through our existing testing processes, and what steps we are taking so that it does not happen again in the future.

## Symptoms of the bug

First, let us look at the scenario in which the bug surfaces:

1. Compile a library A with Scala 3.3.2
2. Compile a project B, depending on A, with Scala 3.3.1 (or 3.3.0)
3. The older compiler crashes with the following message:

  ```
  [error] error while loading SomeClass,
  [error] class file .../SomeClass.class is broken, reading aborted with class java.util.NoSuchElementException
  [error] contextual$
  [error] one error found
  ```

> The error surfaces as soon as a TASTy file from A containing a `using` clause with an inferred name is required to compile project B.

Reviewer: ```suggestion
The error surfaces as soon as a TASTy file from A containing a method with context-function result type is required to compile project B.
```
